fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/HNSW.java

```diff
@@ -0,0 +1,2022 @@
+     * @param storageTransform an affine transformation operator that is used to transform the fetched vector into the
+     *        storage space that is currently being used
+     * @param estimator the estimator to use
+     * @param entryNeighbors A collection of starting nodes for the search in this layer, with their distances
+     * to the query vector already calculated.
+     * @param layer The zero-based index of the layer to search.
+     * @param efSearch The size of the dynamic candidate list. A larger value increases recall at the
+     * cost of performance.
+     * @param nodeCache A cache of nodes that have already been fetched from storage to avoid redundant I/O.
+     * @param queryVector The vector for which to find the nearest neighbors.
+     *
+     * @return A {@link java.util.concurrent.CompletableFuture} that, upon completion, will contain a list of the
+     * best candidate nodes found in this layer, paired with their full node data.
+     */
+    @Nonnull
+    private <N extends NodeReference> CompletableFuture<List<NodeReferenceAndNode<N>>>
+            searchLayer(@Nonnull final StorageAdapter<N> storageAdapter,
+                        @Nonnull final ReadTransaction readTransaction,
+                        @Nonnull final AffineOperator storageTransform,
+                        @Nonnull final Estimator estimator,
+                        @Nonnull final Collection<NodeReferenceWithDistance> entryNeighbors,
+                        final int layer,
+                        final int efSearch,
+                        @Nonnull final Map<Tuple, Node<N>> nodeCache,
+                        @Nonnull final RealVector queryVector) {
+        final Set<Tuple> visited = Sets.newConcurrentHashSet(NodeReference.primaryKeys(entryNeighbors));
+        final Queue<NodeReferenceWithDistance> candidates =
+                new PriorityBlockingQueue<>(config.getM(),
+                        Comparator.comparing(NodeReferenceWithDistance::getDistance));
+        candidates.addAll(entryNeighbors);
+        final Queue<NodeReferenceWithDistance> nearestNeighbors =
+                new PriorityBlockingQueue<>(config.getM(),
+                        Comparator.comparing(NodeReferenceWithDistance::getDistance).reversed());
+        nearestNeighbors.addAll(entryNeighbors);
+
+        return AsyncUtil.whileTrue(() -> {
+            if (candidates.isEmpty()) {
+                return AsyncUtil.READY_FALSE;
+            }
+
+            final NodeReferenceWithDistance candidate = candidates.poll();
+            final NodeReferenceWithDistance furthestNeighbor = Objects.requireNonNull(nearestNeighbors.peek());
+
+            if (candidate.getDistance() > furthestNeighbor.getDistance()) {
+                return AsyncUtil.READY_FALSE;
+            }
+
+            return fetchNodeIfNotCached(storageAdapter, readTransaction, storageTransform, layer, candidate, nodeCache)
```

Comment:
Just to clarify: if I'm following the logic correctly, then `fetchNodeIfNotCached` should do a fetch here if we have `INLINING` nodes, and it should rely on the cache if we have `COMPACT` nodes? Then `fetchNeighborhood` is the reverse--`COMPACT` nodes do fetches of the nodes there, populating the cache, while `INLINING` nodes rely on the fact that it already has the neighborhood vectors